### ISSUE TITLE
Arabic & Hebrew unique issues associated with sections 2.4 and 2.5

### ISSUE THREAD

**Issue by tomerm** [w3c/charmod-norm#80]
I would like to comment on following sections of specification:
     - 2.4 Unicode Controls and Invisible Markers
     - 2.5 Legacy Character Encodings

[ comment from here moved to issue #82 to avoid mixing two differen threads ]

**On legacy characters encoding**
One of legacy code pages which worth considering in this context is https://en.wikipedia.org/wiki/EBCDIC. Majority of Bidi (Arabic / Hebrew) data stored on mainframe systems is in this code page (along with minority stored in Unicode). The problem is that Bidi data stored on modern operating systems (i.e. windows, android, iOS etc.) is radically different from mainframe. Bidi data is stored on those systems in different bidi layouts. Bidi layout basically affects relative order of characters / text segments in the text buffer. Code page conversion not necessarily takes this info account. Thus when data from mainframe is converted to Unicode it can be still in visual bidi layout used on mainframes (instead of logical bidi layout used on modern operating systems). It is impossible to compare data in different bidi layouts. It will lead to incorrect sort / search results.  
The suggestion is to assure that text being sorted / searched are present in the same bidi layout.

**Comment by duerst** [comment]
Just some small comments: EBCDIC isn't a single code page. There are many variants.
My understanding was that legacy (EBCDIC) Hebrew was stored visually, but that Arabic was stored logically from a much earlier point in time, because the need for connecting adjacent characters made it easier to also include bidirectional reordering. This should be checked.

**Comment by tomerm** [comment]
Apologies for being ambiguous. Many of EBCDIC code pages are mentioned at: https://www-01.ibm.com/software/globalization/ccsid/ccsid_registered.html
In my previous comment I referred to 
- EBCDIC Hebrew (cp-424) - https://www-01.ibm.com/software/globalization/ccsid/ccsid424.html and also here: http://www.tachyonsoft.com/cp00424.htm
- EBCDIC Arabic (cp-420) - https://www-01.ibm.com/software/globalization/ccsid/ccsid420.html and also here: http://www.tachyonsoft.com/cp00420.htm

Both Arabic and Hebrew EBCDIC (cp-424 and cp-420) are visual. The visual nature stems from correlation between storage and display on green screens (old ones: https://en.wikipedia.org/wiki/IBM_3270 or screen emulators - modern software which mimic old screens via desktop app such as IBM HOD - http://www-03.ibm.com/software/products/en/rationalhostondemand or non IBM such as ERICOM  - http://www.ericom.com/tn3270.asp ). 

In modern operating systems (i.e. Windows, Android etc.), text stored in the buffer in the chronological order (the same order in which it was typed), undergoes reordering (performed by UBA [Unicode Bidi Algorithm] compliant Bidi engine) before it is rendered on the screen. As opposed to modern operating systems, on legacy systems (i.e. green screens) there is no any transformation which text undergoes before it is rendered. Namely it is rendered exactly like it appears in the storage.

**Comment by aphillips** [comment]
@tomerm Although I appreciate the existence of and difficulties inherent in working with EBCDIC code pages, I would point out that none of these encodings are valid on the Web or in Web standards (where ASCII-compatible encodings rule). This is not meant to trivialize or dismiss your comment, although it does seem orthogonal to our document's purpose.

There are ASCII-compatible visually-ordered encodings of bidirectional scripts, notably the Hebrew encoding ISO-8859-8 (IIRC, on the Web the Arabic encoding ISO 8859-6 is always logically ordered, although I may be mistaken), so this may be a consideration on some level. However, I'm at a loss for how to address this in our context currently.

Can you suggest specific textual amendments?

**Comment by tomerm** [comment]
Code page is not a problem. The problem is in different bidi layouts in which data can be present. Different bidi layouts are realized in different position of characters / text segments. Thus comparing of the same data in different bidi layouts will most definitely produce incorrect results. Since for historic reasons different bidi layouts are associated with EBCDIC code page and legacy systems I mentioned both. However, in general case, even data stored in Unicode can also be present in different Bidi layouts. Bidi layouts and code pages are completely orthogonal concepts (from functional / technical perspectives). 

The suggested textual amendments is as follows:
    assure that text being sorted / searched is present in the same bidi layout.

Normalization to same bidi layout is conceptually similar to code page conversion. Before you can compare two pieces of text you must assure they are encoded with the same code page (i.e. Unicode). Very similarly, if you wish to compare two pieces of Bidi text, you must assure they are transformed to common bidi layout. For more information on bidi layouts please see: http://www.ibm.com/developerworks/websphere/library/techarticles/bidi/bidigen.html

PS. The encodings you mentioned are relevant for display (browser interpret data encoded in such encodings differently). When we are talking about search  / sort, we refer to text in the storage.

**Comment by aphillips** [comment]
@tomerm This still doesn't make sense to me. What do you mean by "layout"? Application of the Unicode Bidirectional Algorithm? 

The encodings I mentioned are character encodings used to _store_ the data. Their relevance to this discussion is that they store data differently from character encodings that store data in logical order: these encodings store the bytes representing each character in "teletype order" (the order that a left-to-right printing line printer would output them). This certainly can make searching more difficult because the source and target text might need to be re-ordered in order to compare them and things such as line-breaking may change the order of the characters in storage. 

If this is what you mean, note that the layout issues don't affect text comparison in this document, since once both the source and target text have been converted to Unicode all of the presentational variation has (hopefully) been resolved (or the text will be muddled).

**Comment by tomerm** [comment]
Some introduction materials on Bidi layout (aka Bidi format if you wish).

Bidirectional text can be stored in several formats (a.k.a. "layouts"), depending on the platform and the intended use. When bidi data is available but not in the appropriate layout, it is possible to convert it to the required layout by performing a "bidi transformation". On most platforms, there exist libraries of tools (often called "bidi engines") to execute bidi transformations.

Bidi text can be stored in a variety of layouts characterized by a combination of 5 attributes:
- ordering scheme
- base direction
- symmetric swapping
- text shaping (for Arabic)
- numeric shaping (for Arabic)

Historically, data stored on mainframes (zSeries and iSeries) was in "visual" layout, i.e. according to a visual ordering scheme, ready for presentation on "dumb" devices like terminals or printers. With the advent of processing power closer to end users, as is the case for workstations and personal computers, it became common to store data in a newer "logical" layout.

Currently, data in visual layout are still preponderant on zSeries and iSeries, while in Windows systems most data are created and processed in logical layout.
Systems running Unix or its variants can handle data in either logical or visual layout, probably with some more affinity for the logical layout.
Java GUI components expect bidi textual data to be in logical layout.
bidi text within HTML may be in either logical or visual layout, but it is generally more convenient to format the data in logical layout, and browsers' support for data in logical layout is also better.

As we have seen, bidi textual data can be stored according to various layouts. Visual layouts can be used straightforwardly for presentation, while logical layouts are more appropriate for most forms of processing.

What happens if some data need processing, but are currently accessible in a visual layout? They must be converted to a logical layout. This operation is not trivial. It entails reversing parts of the text, but not all of it. It entails replacing some occurrences of symmetric symbols by their symmetric match. For Arabic, it also entails choosing the proper shape for each letter and the proper digits for each number. Fortunately, software tools already exist to accomplish such conversions.

In general, the operation which converts a piece of bidi text from a source layout to a target layout is called a "bidi transformation". This can include going from a logical layout to a visual layout, or going the opposite way. The transformation may even leave the ordering scheme unchanged, but alter some or all of the other bidi attributes.

**Comment by tomerm** [comment]
> If this is what you mean, note that the layout issues don't affect text comparison 
> in this document, since once both the source and target text have been 
> converted to Unicode all of the presentational variation has (hopefully) 
> been resolved (or the text will be muddled).

Indeed this is one of the use cases I had in mind. But please observe that just code page conversion to Unicode might not resolve all the issues. Specifically with Bidi text, it must be transformed to logical bidi layout. This is all what this comment is about: **when one converts text to Unicode, he/she must also assure that text is transformed to logical bidi layout.** This unfortunately does not happen automatically (one of the reasons is in fact that bidi layout can't be guessed by just looking at the text Think of palindromes: "swap paws" - are we talking about text running from left to right or from right to left). .

**Comment by tomerm** [comment]
Some more info on bidi layout attributes mentioned above: 

**Base direction** Identifies where in the presentation area bidirectional text starts: right-to-left, left-to-right, or contextual (based on the directionality of the first character that has a defined direction).

**Ordering scheme**  Identifies what kind of algorithm is used when transforming the text layout. There are several types of algorithms:
_Visual algorithms,_ which copy entire lines of text as they are presented without regard for existing embedded directional segments
_Implicit algorithms_
_Logical algorithms,_ which recognize directional segments based on the natural directionality of the characters (right to left for Arabic characters and left to right for English characters), and perform segment inversions accordingly.
_Explicit algorithms,_ which recognize directional segments and perform inversions based on special, explicit, directional controls embedded in the text

**Text shaping**
Indicates whether or not text shaping determination is required when text is rendered. This is relevant for Arabic scripts, since characters assume different shapes (initial, medial, final, or isolated) according to their position in a word and the connectivity traits of the surrounding characters.

**Symmetric swapping**
In bidi text, this Indicates whether or not specific directional pairs of characters, such as left and right parentheses, and greater-than and less-than signs, will have to be interchanged in order to preserve the logical meaning of the presented text

**Numeral shapes**
Indicates whether or not numbers embedded in an Arabic script will have to be presented using the Arabic (or European) digit shapes or the Hindi (or Arabic-Indic) digit shapes.

**Comment by aphillips** [comment]
@tomerm Thanks for this. I'm well aware of these considerations. Under separate cover I asked Mati Allouche to make suggestions and I have adapted his suggestion in a new paragraph in Section 2.5 (Legacy Character Encodings). Note that I also added a reference to Section 3.3.1 of Charmod (the original Charmod) which provides illustrations, terminology, and so forth.

**Comment by tomerm** [comment]
Thanks. I looked at the addition to section 2.5 at: http://w3c.github.io/charmod-norm/#legacyCharacterEncoding 
_"..One additional consideration in converting to Unicode is the existence of legacy character encodings for bidirectional languages that use a visual storage order. That is, unlike Unicode and other modern encodings, the characters are stored in memory in the order that they are printed on the screen from left-to-right (as with a line printer). When converting these encodings to Unicode or when comparing text in these encodings, care must be taken to place both the source and target text into logical order. For more information, see Section 3.3.1 of [CHARMOD]..."_

Overall looks very good. Thanks. Just one minor comment: _".. for bidirectional languages..."_ better to say  **for languages with bidirectional scripts (such as Arabic or Hebrew)**
This should be also a position of Mati (just for your awareness - we worked together for more than 10 years in IBM in the same field. When Mati retired I stepped into his big shoes).
Appreciate your attention to my comments.
